src/renderers/native/ReactNativeViewConfigRegistry.js

```diff
@@ -14,29 +14,53 @@
 
 const invariant = require('fbjs/lib/invariant');
 
-export type ReactNativeBaseComponentViewConfig = {
-  validAttributes: Object,
-  uiViewClassName: string,
-  propTypes?: Object,
-};
+import type {
+  ReactNativeBaseComponentViewConfig,
+  ViewConfigGetter,
+} from 'ReactNativeTypes';
 
+const viewConfigCallbacks = new Map();
 const viewConfigs = new Map();
 
 const ReactNativeViewConfigRegistry = {
-  register(viewConfig: ReactNativeBaseComponentViewConfig) {
-    const name = viewConfig.uiViewClassName;
+  /**
+   * Registers a native view/component by name.
+   * A callback is provided to load the view config from UIManager.
+   * The callback is deferred until the view is actually rendered.
+   * This is done to avoid causing Prepack deopts.
+   */
+  register(name: string, callback: ViewConfigGetter): string {
     invariant(
-      !viewConfigs.has(name),
+      !viewConfigCallbacks.has(name),
       'Tried to register two views with the same name %s',
       name,
     );
-    viewConfigs.set(name, viewConfig);
+    viewConfigCallbacks.set(name, callback);
     return name;
   },
-  get(name: string) {
-    const config = viewConfigs.get(name);
-    invariant(config, 'View config not found for name %s', name);
-    return config;
+
+  /**
+   * Retrieves a config for the specified view.
+   * If this is the first time the view has been used,
+   * This configuration will be lazy-loaded from UIManager.
+   */
+  get(name: string): ReactNativeBaseComponentViewConfig {
+    let viewConfig;
+    if (!viewConfigs.has(name)) {
+      const callback = viewConfigCallbacks.get(name);
+      invariant(
+        typeof callback === 'function',
+        'View config not found for name %s',
+        name,
+      );
+      viewConfigCallbacks.set(name, null);
```

Comment:
Any reason this isn't .delete? Either way.

Reply:
I use `viewConfigCallbacks` to check for view name uniqueness. Nulling out prevents us from blocking GC while avoiding having to check both `viewConfigs` and `viewConfigCallbacks` maps.

Reply:
:+1: